File: Object-C/BlackFireIM/Example/Lib/MSIMSDK-UIKit/ChatDetail/Controller/MSInputViewController.m
//
//  MSInputViewController.m
//  BlackFireIM
//
//  Created by benny wang on 2021/3/4.
//

#import "MSInputViewController.h"
#import "MSIMSDK-UIKit.h"
#import "MSFaceUtil.h"

typedef NS_ENUM(NSUInteger, InputStatus) {
    Input_Status_Input,
    Input_Status_Input_Face,
    Input_Status_Input_More,
    Input_Status_Input_Keyboard,
    Input_Status_Input_Talk,
};
@interface MSInputViewController ()<MSInputBarViewDelegate,MSChatMoreViewDelegate,MSFaceViewDelegate,MSMenuViewDelegate>

@property (nonatomic, assign) InputStatus status;

@end

@implementation MSInputViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self setupUI];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillHide:) name:UIKeyboardWillHideNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillShow:) name:UIKeyboardWillShowNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillChangeFrame:) name:UIKeyboardWillChangeFrameNotification object:nil];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)dealloc
{
    MSLog(@"%@ dealloc",self.class);
}

- (void)setupUI
{
    self.view.backgroundColor = [UIColor d_colorWithColorLight:TInput_Background_Color dark:TInput_Background_Color_Dark];
    _status = Input_Status_Input;
    _inputBar = [[MSInputBarView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, TTextView_Height)];
    _inputBar.delegate = self;
    [self.view addSubview:_inputBar];
}

- (void)keyboardWillHide:(NSNotification *)notification
{
    // http://tapd.oa.com/20398462/bugtrace/bugs/view?bug_id=1020398462072883317&url_cache_key=b8dc0f6bee40dMSe0e702ef8cebd5d81
    if (_delegate && [_delegate respondsToSelector:@selector(inputController:didChangeHeight:)]){
        [_delegate inputController:self didChangeHeight:_inputBar.frame.size.height + Bottom_SafeHeight];
    }
}

- (void)keyboardWillShow:(NSNotification *)notification
{
    if(_status == Input_Status_Input_Face){
        [self hideFaceAnimation];
    }else if(_status == Input_Status_Input_More){
        [self hideMoreAnimation];
    }else{
        //[self hideFaceAnimation:NO];
        //[self hideMoreAnimation:NO];
    }
    _status = Input_Status_Input_Keyboard;
}

- (void)keyboardWillChangeFrame:(NSNotification *)notification
{
    CGRect keyboardFrame = [notification.userInfo[UIKeyboardFrameEndUserInfoKey] CGRectValue];
    if (_delegate && [_delegate respondsToSelector:@selector(inputController:didChangeHeight:)]){
        [_delegate inputController:self didChangeHeight:keyboardFrame.size.height + _inputBar.frame.size.height];
    }
}

- (void)hideFaceAnimation
{
    self.faceView.hidden = NO;
    self.faceView.alpha = 1.0;
    self.menuView.hidden = NO;
    self.menuView.alpha = 1.0;
    __weak typeof(self) ws = self;
    [UIView animateWithDuration:0.3 delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^{
        ws.faceView.alpha = 0.0;
        ws.menuView.alpha = 0.0;
    } completion:^(BOOL finished) {
        ws.faceView.hidden = YES;
        ws.faceView.alpha = 1.0;
        ws.menuView.hidden = YES;
        ws.menuView.alpha = 1.0;
        [ws.menuView removeFromSuperview];
        [ws.faceView removeFromSuperview];
    }];
}

- (void)showFaceAnimation
{
    [self.view addSubview:self.faceView];
    [self.view addSubview:self.menuView];

    self.faceView.hidden = NO;
    CGRect frame = self.faceView.frame;
    frame.origin.y = Screen_Height;
    self.faceView.frame = frame;

    self.menuView.hidden = NO;
    frame = self.menuView.frame;
    frame.origin.y = self.faceView.frame.origin.y + self.faceView.frame.size.height;
    self.menuView.frame = frame;

    __weak typeof(self) ws = self;
    [UIView animateWithDuration:0.3 delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^{
        CGRect newFrame = ws.faceView.frame;
        newFrame.origin.y = ws.inputBar.frame.origin.y + ws.inputBar.frame.size.height;
        ws.faceView.frame = newFrame;

        newFrame = ws.menuView.frame;
        newFrame.origin.y = ws.faceView.frame.origin.y + ws.faceView.frame.size.height;
        ws.menuView.frame = newFrame;
    } completion:nil];
}

- (void)hideMoreAnimation
{
    self.moreView.hidden = NO;
    self.moreView.alpha = 1.0;
    __weak typeof(self) ws = self;
    [UIView animateWithDuration:0.3 delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^{
        ws.moreView.alpha = 0.0;
    } completion:^(BOOL finished) {
        ws.moreView.hidden = YES;
        ws.moreView.alpha = 1.0;
        [ws.moreView removeFromSuperview];
    }];
}

- (void)showMoreAnimation
{
    [self.view addSubview:self.moreView];

    self.moreView.hidden = NO;
    CGRect frame = self.moreView.frame;
    frame.origin.y = Screen_Height;
    self.moreView.frame = frame;
    __weak typeof(self) ws = self;
    [UIView animateWithDuration:0.3 delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^{
        CGRect newFrame = ws.moreView.frame;
        newFrame.origin.y = ws.inputBar.frame.origin.y + ws.inputBar.frame.size.height;
        ws.moreView.frame = newFrame;
    } completion:nil];
}

#pragma mark MSInputBarViewDelegate
- (void)inputBarDidTouchVoice:(MSInputBarView *)textView
{
    if(_status == Input_Status_Input_Talk){
        return;
    }
    [_inputBar.inputTextView resignFirstResponder];
    [self hideFaceAnimation];
    [self hideMoreAnimation];
    _status = Input_Status_Input_Talk;
    if (_delegate && [_delegate respondsToSelector:@selector(inputController:didChangeHeight:)]){
        [_delegate inputController:self didChangeHeight:TTextView_Height + Bottom_SafeHeight];
    }
}

- (void)inputBarDidTouchMore:(MSInputBarView *)textView
{
    if(_status == Input_Status_Input_More){
        return;
    }
    if(_status == Input_Status_Input_Face){
        [self hideFaceAnimation];
    }
    [_inputBar.inputTextView resignFirstResponder];
    [self showMoreAnimation];
    _status = Input_Status_Input_More;
    if (_delegate && [_delegate respondsToSelector:@selector(inputController:didChangeHeight:)]){
        [_delegate inputController:self didChangeHeight:_inputBar.frame.size.height + self.moreView.frame.size.height + Bottom_SafeHeight];
    }
}

- (void)inputBarDidTouchFace:(MSInputBarView *)textView
{
    if(_status == Input_Status_Input_More){
        [self hideMoreAnimation];
    }
    [_inputBar.inputTextView resignFirstResponder];
    [self showFaceAnimation];
    _status = Input_Status_Input_Face;
    if (_delegate && [_delegate respondsToSelector:@selector(inputController:didChangeHeight:)]){
        [_delegate inputController:self didChangeHeight:_inputBar.frame.size.height + self.faceView.frame.size.height + self.menuView.frame.size.height + Bottom_SafeHeight];
    }
}

- (void)inputBarDidTouchKeyboard:(MSInputBarView *)textView
{
    if(_status == Input_Status_Input_More){
        [self hideMoreAnimation];
    }
    if (_status == Input_Status_Input_Face) {
        [self hideFaceAnimation];
    }
    _status = Input_Status_Input_Keyboard;
    [_inputBar.inputTextView becomeFirstResponder];
}

- (void)inputBar:(MSInputBarView *)textView didChangeInputHeight:(CGFloat)offset
{
    if(_status == Input_Status_Input_Face){
        [self showFaceAnimation];
    }
    else if(_status == Input_Status_Input_More){
        [self showMoreAnimation];
    }
    if (_delegate && [_delegate respondsToSelector:@selector(inputController:didChangeHeight:)]){
        [_delegate inputController:self didChangeHeight:self.view.frame.size.height + offset];
    }
}

- (void)inputBar:(MSInputBarView *)textView didSendText:(NSString *)text
{
    if(_delegate && [_delegate respondsToSelector:@selector(inputController:didSendTextMessage:)]){
        [_delegate inputController:self didSendTextMessage:text];
    }
}

- (void)inputBar:(MSInputBarView *)textView didSendVoice:(NSString *)path
{
    if(_delegate && [_delegate respondsToSelector:@selector(inputController:didSendVoiceMessage:)]){
        [_delegate inputController:self didSendVoiceMessage:path];
    }
}

- (void)inputBarDidInputAt:(MSInputBarView *)textView
{
    if (_delegate && [_delegate respondsToSelector:@selector(inputControllerDidInputAt:)]) {
        [_delegate inputControllerDidInputAt:self];
    }
}

- (void)inputBar:(MSInputBarView *)textView didDeleteAt:(NSString *)atText
{
    if (_delegate && [_delegate respondsToSelector:@selector(inputController:didDeleteAt:)]) {
        [_delegate inputController:self didDeleteAt:atText];
    }
}

- (void)inputBarContentDidChanged:(MSInputBarView *_Nonnull)textView
{
    if (_delegate && [_delegate respondsToSelector:@selector(inputController:contentDidChanged:)]) {
        [_delegate inputController:self contentDidChanged:textView.inputTextView.text];
    }
}

- (void)reset
{
    if(_status == Input_Status_Input){
        return;
    }
    else if(_status == Input_Status_Input_More){
        [self hideMoreAnimation];
    }
    else if(_status == Input_Status_Input_Face){
        [self hideFaceAnimation];
    }
    _status = Input_Status_Input;
    [_inputBar.inputTextView resignFirstResponder];
    if (_delegate && [_delegate respondsToSelector:@selector(inputController:didChangeHeight:)]){
        [_delegate inputController:self didChangeHeight:_inputBar.frame.size.height + Bottom_SafeHeight];
    }
}

- (MSChatMoreView *)moreView
{
    if(!_moreView){
        _moreView = [[MSChatMoreView alloc] initWithFrame:CGRectMake(0, _inputBar.frame.origin.y + _inputBar.frame.size.height, self.view.frame.size.width, 0)];
        _moreView.delegate = self;
        MSInputMoreCellData *cameraData = [[MSInputMoreCellData alloc]initWithType:MSIM_MORE_PHOTO];
        cameraData.title = TUILocalizableString(TUIKitMorePhoto);
        cameraData.image = [UIImage bf_imageNamed:@"more_picture"];
        
        MSInputMoreCellData *photoData = [[MSInputMoreCellData alloc]initWithType:MSIM_MORE_VIDEO];
        photoData.title = TUILocalizableString(TUIKitMoreCamera);
        photoData.image = [UIImage bf_imageNamed:@"more_camera"];
        
//        MSInputMoreCellData *voiceData = [[MSInputMoreCellData alloc]initWithType:MSIM_MORE_VOICE_CALL];
//        voiceData.title = TUILocalizableString(TUIKitMoreVoiceCall);
//        voiceData.image = [UIImage bf_imageNamed:@"more_voice_call"];
//
//        MSInputMoreCellData *videoData = [[MSInputMoreCellData alloc]initWithType:MSIM_MORE_VIDEO_CALL];
//        videoData.title = TUILocalizableString(TUIKitMoreVideoCall);
//        videoData.image = [UIImage bf_imageNamed:@"more_video_call"];
//
        [_moreView setData:@[cameraData,photoData]];
    }
    return _moreView;
}

- (MSFaceView *)faceView
{
    if(!_faceView){
        _faceView = [[MSFaceView alloc] initWithFrame:CGRectMake(0, _inputBar.frame.origin.y + _inputBar.frame.size.height, self.view.frame.size.width, 180)];
        _faceView.delegate = self;
        [_faceView setData:[MSFaceUtil defaultConfig].defaultFace];
    }
    return _faceView;
}

- (MSMenuView *)menuView
{
    if(!_menuView){
        _menuView = [[MSMenuView alloc] initWithFrame:CGRectMake(0, self.faceView.frame.origin.y + self.faceView.frame.size.height, self.view.frame.size.width, 40)];
        _menuView.delegate = self;
    }
    return _menuView;
}

#pragma mark - more view delegate
- (void)moreView:(MSChatMoreView *)moreView didSelectMoreCell:(MSInputMoreCell *)cell
{
    if(_delegate && [_delegate respondsToSelector:@selector(inputController:didSelectMoreCell:)]){
        [_delegate inputController:self didSelectMoreCell:cell];
    }
}

#pragma mark - MSFaceViewDelegate

- (void)faceView:(MSFaceView *)faceView scrollToFaceGroupIndex:(NSInteger)index
{
//    [self.menuView scrollToMenuIndex:index];
}

- (void)faceViewDidBackDelete:(MSFaceView *)faceView
{
    [_inputBar backDelete];
}

- (void)faceView:(MSFaceView *)faceView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    BFFaceGroup *group = [MSFaceUtil defaultConfig].defaultFace[indexPath.section];
    BFFaceCellData *face = group.faces[indexPath.row];
    if(indexPath.section == 0){
        NSString *faceName = [face.name substringFromIndex:@"emoji/".length];
        [_inputBar addEmoji:faceName];
    }else{
        //直接发送
        // to do
    }
}

#pragma mark - MSMenuViewDelegate

- (void)menuViewDidSendMessage:(MSMenuView *)menuView
{
    NSString *text = [_inputBar getInput];
    if([text isEqualToString:@""]){
        return;
    }
    [_inputBar clearInput];
    if(_delegate && [_delegate respondsToSelector:@selector(inputController:didSendTextMessage:)]){
        [_delegate inputController:self didSendTextMessage:text];
    }
}

@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/ChatDetail/Controller/MSInputViewController.swift
//
//  MSInputViewController.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/8/30.
//

import UIKit
import MSIMSDK


public protocol MSInputViewControllerDelegate: NSObjectProtocol {
    
    ///当前 InputController 高度改变时的回调
    func didChangeHeight(inputController: MSInputViewController,height: CGFloat)
    ///当前 InputCOntroller 发送文本消息时的回调
    func didSendTextMessage(inputController: MSInputViewController,msg: String)
    ///当前 InputCOntroller 发送语音信息时的回调
    func didSendVoiceMessage(inputController: MSInputViewController,filePath: String)
    ///输入框中内容发生变化时的回调
    func contentDidChanged(inputController: MSInputViewController,text: String)
    ///有 @ 字符输入
    func inputControllerDidInputAt(inputController: MSInputViewController)
    ///有 @xxx 字符删除
    func didDeleteAt(inputController: MSInputViewController,atText: String)
    ///点击拍照，照片等更多功能
    func didSelectMoreCell(inputController: MSInputViewController,cell: MSInputMoreCell)
    /// 点击发送自定义表情
    func didSendEmotion(inputController: MSInputViewController,data: MSFaceCellData)
}

enum InputStatus {
    case input,face,more,keyboard,talk
}
public class MSInputViewController: UIViewController {

    var inputBar: MSInputBarView!
    
    lazy var faceView: MSFaceView = {
        let faceView = MSFaceView(frame: CGRect(x: 0, y: inputBar.top + inputBar.height, width: view.width, height: 180))
        faceView.delegate = self
        faceView.setData(data: MSFaceUtil.shared.faceGroups)
        return faceView
    }()
    
    lazy var menuView: MSMenuView = {
        let menuView = MSMenuView(frame: CGRect(x: 0, y: faceView.top + faceView.height, width: view.width, height: 40))
        menuView.delegate = self
        var menus: [MSMenuCellData] = []
        for (index,group) in MSFaceUtil.shared.faceGroups.enumerated() {
            let data = MSMenuCellData()
            data.normalPath = group.menuNormalPath
            data.selectPath = group.menuSelectPath
            data.isSelected = index == 0
            menus.append(data)
        }
        menuView.data = menus
        return menuView
    }()
    
    lazy var moreView: MSChatMoreView = {
        let moreView = MSChatMoreView(frame: CGRect(x: 0, y: inputBar.top + inputBar.height, width: view.width, height: 0))
        moreView.delegate = self
        var cameraData = MSInputMoreCellData(type: .photo)
        cameraData.title = Bundle.bf_localizedString(key: "TUIKitMorePhoto")
        cameraData.image = UIImage.bf_imageNamed(name: "more_picture")
        
        var photoData = MSInputMoreCellData(type: .video)
        photoData.title = Bundle.bf_localizedString(key: "TUIKitMoreVideo")
        photoData.image = UIImage.bf_imageNamed(name: "more_camera")
        
        var voiceData = MSInputMoreCellData(type: .voiceCall)
        voiceData.title = Bundle.bf_localizedString(key: "TUIKitMoreVoiceCall")
        voiceData.image = UIImage.bf_imageNamed(name: "more_voice_call")
        
        var videoData = MSInputMoreCellData(type: .videoCall)
        videoData.title = Bundle.bf_localizedString(key: "TUIKitMoreVideoCall")
        videoData.image = UIImage.bf_imageNamed(name: "more_video_call")
        
        var locationData = MSInputMoreCellData(type: .location)
        locationData.title = Bundle.bf_localizedString(key: "TUIKitMoreLocation")
        locationData.image = UIImage.bf_imageNamed(name: "more_location")
        
        if self.type == .MSIM_CHAT_TYPE_C2C {
            moreView.setData(data: [cameraData,photoData,voiceData,videoData,locationData])
        }else {
            moreView.setData(data: [cameraData,photoData])
        }
        return moreView
    }()
    
    private weak var delegate: MSInputViewControllerDelegate?
    
    private var type: MSIMAChatType = .MSIM_CHAT_TYPE_C2C
    
    private var status: InputStatus = .input
    
    public func reset() {
        
        if status == .input {return}
        if status == .more {
            hideMoreAnimation()
        }else if status == .face {
            hideFaceAnimation()
        }
        status = .input
        inputBar.inputTextView.resignFirstResponder()
        delegate?.didChangeHeight(inputController: self, height: inputBar.height + UIScreen.safeAreaBottomHeight)
    }
    
    init(chatType: MSIMAChatType,delegate: MSInputViewControllerDelegate?) {
        super.init(nibName: nil, bundle: nil)
        self.type = chatType
        self.delegate = delegate
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    deinit {
        print("\(self) dealloc")
    }
    
    public override func viewDidLoad() {
        super.viewDidLoad()

        setupUI()
    }
    
    public override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillHide), name: UIResponder.keyboardWillHideNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillShow), name: UIResponder.keyboardWillShowNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillChangeFrame), name: UIResponder.keyboardWillChangeFrameNotification, object: nil)
    }
    
    public override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        NotificationCenter.default.removeObserver(self)
    }
    
    private func setupUI() {
        view.backgroundColor = UIColor.d_color(light: MSMcros.TInput_Background_Color, dark: MSMcros.TInput_Background_Color_Dark)
        inputBar = MSInputBarView(frame: CGRect(x: 0, y: 0, width: view.width, height: MSMcros.TTextView_Height))
        inputBar.delegate = self
        view.addSubview(inputBar)
    }
}

extension MSInputViewController {
    
    @objc func keyboardWillHide() {
        delegate?.didChangeHeight(inputController: self, height: inputBar.height + UIScreen.safeAreaBottomHeight)
    }
    
    @objc func keyboardWillShow() {
        
        if status == .face {
            hideFaceAnimation()
        }else if status == .more {
            hideMoreAnimation()
        }else {
            
        }
        status = .keyboard
    }
    
    @objc func keyboardWillChangeFrame(note: Notification) {
        if let keyboardFrame = note.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect {
            delegate?.didChangeHeight(inputController: self, height: keyboardFrame.height + inputBar.height)
        }
    }
    
    func hideFaceAnimation() {
        faceView.isHidden = false
        faceView.alpha = 1
        menuView.isHidden = false
        menuView.alpha = 1
        UIView.animate(withDuration: 0.3, delay: 0, options: .curveEaseOut) {[weak self] in
            self?.faceView.alpha = 0
            self?.menuView.alpha = 0
        } completion: {[weak self] _ in
            self?.faceView.isHidden = true
            self?.faceView.alpha = 1
            self?.menuView.isHidden = true
            self?.menuView.alpha = 1
            self?.menuView.removeFromSuperview()
            self?.faceView.removeFromSuperview()
        }
    }
    
    func showFaceAnimation() {
        view.addSubview(faceView)
        view.addSubview(menuView)
        
        faceView.isHidden = false
        var frame = faceView.frame
        frame.origin.y = UIScreen.height
        faceView.frame = frame
        
        menuView.isHidden = false
        frame = menuView.frame
        frame.origin.y = faceView.top + faceView.height
        menuView.frame = frame
        
        UIView.animate(withDuration: 0.3, delay: 0, options: .curveEaseOut) {
            
            var newFrame = self.faceView.frame
            newFrame.origin.y = self.inputBar.top + self.inputBar.height
            self.faceView.frame = newFrame
            
            newFrame = self.menuView.frame
            newFrame.origin.y = self.faceView.top + self.faceView.height
            self.menuView.frame = newFrame
            
        } completion: { _ in
            
        }
    }
    
    func hideMoreAnimation() {
        
        moreView.isHidden = false
        moreView.alpha = 1
        UIView.animate(withDuration: 0.3, delay: 0, options: .curveEaseOut) {
            self.moreView.alpha = 0
        } completion: { _ in
            
            self.moreView.isHidden = true
            self.moreView.alpha = 1
            self.moreView.removeFromSuperview()
        }
    }
    
    func showMoreAnimation() {
        view.addSubview(moreView)
        moreView.isHidden = false
        var frame = moreView.frame
        frame.origin.y = UIScreen.height
        moreView.frame = frame
        UIView.animate(withDuration: 0.3, delay: 0, options: .curveEaseOut) {
            
            var newFrame = self.moreView.frame
            newFrame.origin.y = self.inputBar.top + self.inputBar.height
            self.moreView.frame = newFrame
            
        } completion: { _ in
            
        }
    }
}

extension MSInputViewController: MSInputBarViewDelegate {
    public func inputBarDidTouchFace(textView: MSInputBarView) {
        if status == .more {
            hideMoreAnimation()
        }
        inputBar.inputTextView.resignFirstResponder()
        showFaceAnimation()
        status = .face
        delegate?.didChangeHeight(inputController: self, height: inputBar.height + faceView.height + menuView.height + UIScreen.safeAreaBottomHeight)
    }
    
    public func inputBarDidTouchMore(textView: MSInputBarView) {
        if status == .more {
            return
        }
        if status == .face {
            hideFaceAnimation()
        }
        inputBar.inputTextView.resignFirstResponder()
        showMoreAnimation()
        status = .more
        delegate?.didChangeHeight(inputController: self, height: inputBar.height + moreView.height + UIScreen.safeAreaBottomHeight)
    }
    
    public func inputBarDidTouchVoice(textView: MSInputBarView) {
        
        if status == .talk {
            return
        }
        inputBar.inputTextView.resignFirstResponder()
        hideFaceAnimation()
        hideMoreAnimation()
        status = .talk
        delegate?.didChangeHeight(inputController: self, height: MSMcros.TTextView_Height + UIScreen.safeAreaBottomHeight)
    }
    
    public func inputBarDidChangeInputHeight(textView: MSInputBarView, offset: CGFloat) {
        
        if status == .face {
            showFaceAnimation()
        }else if status == .more {
            showMoreAnimation()
        }
        delegate?.didChangeHeight(inputController: self, height: view.height + offset)
    }
    
    public func inputBarDidSendText(textView: MSInputBarView, text: String) {
        delegate?.didSendTextMessage(inputController: self, msg: text)
    }
    
    public func inputBarDidSendVoice(textView: MSInputBarView, path: String) {
        delegate?.didSendVoiceMessage(inputController: self, filePath: path)
    }
    
    public func inputBarDidInputAt(textView: MSInputBarView) {
        delegate?.inputControllerDidInputAt(inputController: self)
    }
    
    public func inputBarDidDeleteAt(textView: MSInputBarView, text: String) {
        delegate?.didDeleteAt(inputController: self, atText:text )
    }
    
    public func inputBarDidTouchKeyboard(textView: MSInputBarView) {
        
        if status == .more {
            hideMoreAnimation()
        }
        if status == .face {
            hideFaceAnimation()
        }
        status = .keyboard
        inputBar.inputTextView.becomeFirstResponder()
    }
    
    public func inputBarContentDidChanged(textView: MSInputBarView) {
        
        delegate?.contentDidChanged(inputController: self, text: textView.inputTextView.text)
    }
    
}


extension MSInputViewController: MSChatMoreViewDelegate,MSFaceViewDelegate,MSMenuViewDelegate {
    
    public func faceViewDidBackDelete(faceView: MSFaceView) {
        inputBar.backDelete()
    }
    
    
    public func didSelectMoreCell(moreView: MSChatMoreView, cell: MSInputMoreCell) {
        delegate?.didSelectMoreCell(inputController: self, cell: cell)
    }
    
    public func scrollToFaceGroup(faceVeiw: MSFaceView, index: Int) {
        menuView.scrollToMenu(at: index, in: faceVeiw.faceGroups[index])
    }
    
    public func didSelectItem(faceView: MSFaceView, indexPath: IndexPath) {
        let group = MSFaceUtil.shared.faceGroups[indexPath.section]
        let face = group.faces[indexPath.row]
        if indexPath.section == 0 {
            inputBar.addEmoji(emoji: face.name!)
        }else {
            self.delegate?.didSendEmotion(inputController: self, data: face)
        }
    }
    
    func menuViewDidSendMessage(menuView: MSMenuView) {
        
        let text = inputBar.getInput()
        if text.count == 0 {return}
        inputBar.clearInput()
        delegate?.didSendTextMessage(inputController: self, msg: text)
    }
    
    func menuViewDidSelectItemAtIndex(index: Int) {
        
        self.faceView.scrollToFaceGroup(index: index)
    }
}
